[Fonts] systemFallbackForCharacters() only works on individual characters, not clusters

litherum opened the pull request:
#### 9df8edf812b7c4330d6f9c1578ef33e57f37457c
<pre>
[Fonts] systemFallbackForCharacters() only works on individual characters, not clusters
<a href="https://bugs.webkit.org/show_bug.cgi?id=260758">https://bugs.webkit.org/show_bug.cgi?id=260758</a>
rdar://114481597

Reviewed by Cameron McCormack.

When you pick a font, you have to pick a font for a whole cluster; different codepoints
in a cluster can&apos;t fall back to different fonts. If you&apos;ve exhausted the whole font-family
list and you fall off the end, you need to ask the system for a font that can render
the cluster. The Core Text platform facilities have support for this: they accept a
character pointer and a length. However, the only problem is that our own system fallback
code operated on an individual UChar32, rather than on a whole cluster.

This is a classic case of impedence mismatch: At the beginning of the procedure, you have
clusters, and at the end of the procedure (when we hit Core Text), we operate on clusters
too. This patch simply hooks up cluster support (via StringViews) throughout the interior
of the procedure.

Right now, this patch has no behavior change, because the only codepath that uses this
functionality is our fast text codepath, which currently only operates on a single
character at a time (another way to think of this: every cluster has exactly one code
point inside it). If we want to stop using the complex text codepath, this will be relaxed
in the future, to allow the fast text codepath to operate on whole clusters. When that day
comes, we&apos;ll need to be able to feed the clusters to systemFallbackForCharacters(), like
this patch allows for. (So, right now, this patch doesn&apos;t add any tests, because there is
no behavior change.)

* Source/WebCore/platform/graphics/Font.cpp:
(WebCore::Font::systemFallbackFontForCharacter const):
* Source/WebCore/platform/graphics/Font.h:
* Source/WebCore/platform/graphics/FontCache.h:
* Source/WebCore/platform/graphics/FontCascadeFonts.cpp:
(WebCore::FontCascadeFonts::glyphDataForSystemFallback):
* Source/WebCore/platform/graphics/SystemFallbackFontCache.cpp:
(WebCore::SystemFallbackFontCache::systemFallbackFontForCharacter):
* Source/WebCore/platform/graphics/SystemFallbackFontCache.h:
(WebCore::add):
(WebCore::SystemFallbackFontCache::CharacterFallbackMapKeyHashTraits::constructDeletedValue):
(WebCore::SystemFallbackFontCache::CharacterFallbackMapKeyHashTraits::isDeletedValue):
* Source/WebCore/platform/graphics/cocoa/FontCacheCoreText.cpp:
(WebCore::lookupFallbackFont):
(WebCore::FontCache::systemFallbackForCharacters):
(WebCore::createFontForCharacters): Deleted.
* Source/WebCore/platform/graphics/coretext/FontCascadeCoreText.cpp:

Canonical link: <a href="https://commits.webkit.org/267335@main">https://commits.webkit.org/267335@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/de49d96300bd64364b6b7af0ec8368f8c43f8fbb

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/16322 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/16641 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/17068 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/18091 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/15310 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/16512 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/19736 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/16768 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/17687 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/16517 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/16956 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/14043 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/18858 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/14201 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/14786 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/21591 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/15188 "Passed tests") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/14950 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/18752 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/15537 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/13182 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/14754 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/3903 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/19121 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/15364 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

heycam commented on Source/WTF/wtf/Hasher.h:
```diff
@@ -123,7 +123,7 @@ inline void add(Hasher& hasher, const String& string)
 inline void add(Hasher& hasher, const AtomString& string)
 {
-    // Chose to hash the pointer here. Assuming this is better than hashing the characters or hashing the already-computed hash of the characters.
+    // We are assuming that hashing the characters is better than hashing the possibly-already-computed hash of the characters.
```
We are hashing the AtomStringImpl pointer, so we should retain the meaning of the old comment.

heycam replied:
But this and the WTFString.h change should really be in a separate PR.

litherum replied:
Yeah, reading this again, I totally changed the meaning, which was unintentional. Whoops! This is what reviews are for :)

heycam commented on Source/WebCore/platform/graphics/FontCascadeFonts.cpp:
```diff
@@ -351,7 +351,9 @@ GlyphData FontCascadeFonts::glyphDataForSystemFallback(UChar32 character, const
-    auto systemFallbackFont = font->systemFallbackFontForCharacter(character, description, resolvedEmojiPolicy, m_isForPlatformFont ? IsForPlatformFont::Yes : IsForPlatformFont::No);
+    StringBuilder stringBuilder;
+    stringBuilder.appendCharacter(character);
+    auto systemFallbackFont = font->systemFallbackFontForCharacter(stringBuilder, description, resolvedEmojiPolicy, m_isForPlatformFont ? IsForPlatformFont::Yes : IsForPlatformFont::No);
```
I assume you will append more things to the StringBuilder in a future PR. Otherwise you could probably just write:

```
font->systemFallbackFontForCharacter(makeString(character), description, ...
```

litherum replied:
yeah, good point!

litherum replied:
Turns out `makeString(UChar32)` doesn't exist, so it ends up calling `makeString(UChar)` instead, which resulted in the failures :/

For this patch, it probably doesn't make sense to add an overload, as that's a pretty different thing... I'll put this `StringBuilder` thing back for this patch and I'll try to add an overload (and use it, if possible) in a follow-up patch.

heycam commented on Source/WebCore/platform/graphics/Font.cpp:
```diff
@@ -551,9 +551,9 @@ GlyphBufferAdvance Font::applyTransforms(GlyphBuffer&, unsigned, unsigned, bool,
 {
-    return SystemFallbackFontCache::forCurrentThread().systemFallbackFontForCharacter(this, character, description, resolvedEmojiPolicy, isForPlatformFont);
+    return SystemFallbackFontCache::forCurrentThread().systemFallbackFontForCharacter(this, stringView, description, resolvedEmojiPolicy, isForPlatformFont);
 }
```
Can we use a more descriptive argument name than "stringView" here, and in the other functions touched in this PR? Some of the functions are expecting a single character (so, one 8 bit or one or two 16 bit code units) and others are expecting any number of characters. The functions currently use "character" and "characters" for these, which seem better to me.

I guess in a future PR some or all of these that are expecting a single character will be changing to expect a cluster (comprising one or more code units), at which point "cluster" might be a better name than "character".

Also: for functions expecting a single character (and in the future, a single cluster), should we have assertions on the expected length of the StringView, e.g. that is contains one code point (and in the future, one grapheme cluster)?

litherum replied:
Well, if a function expect a single code point, we'd pass it a `UChar32` like how `glyphDataForSystemFallback()` does. I've updated the functions which take a `StringView` to be `characterCluster`.

Grapheme clusters can be arbitrarily long, so it's probably impossible to make a length assertion.

litherum committed:
[Fonts] systemFallbackForCharacters() only works on individual characters, not clusters
https://bugs.webkit.org/show_bug.cgi?id=260758
rdar://114481597

Reviewed by Cameron McCormack.

When you pick a font, you have to pick a font for a whole cluster; different codepoints
in a cluster can't fall back to different fonts. If you've exhausted the whole font-family
list and you fall off the end, you need to ask the system for a font that can render
the cluster. The Core Text platform facilities have support for this: they accept a
character pointer and a length. However, the only problem is that our own system fallback
code operated on an individual UChar32, rather than on a whole cluster.

This is a classic case of impedence mismatch: At the beginning of the procedure, you have
clusters, and at the end of the procedure (when we hit Core Text), we operate on clusters
too. This patch simply hooks up cluster support (via StringViews) throughout the interior
of the procedure.

Right now, this patch has no behavior change, because the only codepath that uses this
functionality is our fast text codepath, which currently only operates on a single
character at a time (another way to think of this: every cluster has exactly one code
point inside it). If we want to stop using the complex text codepath, this will be relaxed
in the future, to allow the fast text codepath to operate on whole clusters. When that day
comes, we'll need to be able to feed the clusters to systemFallbackForCharacters(), like
this patch allows for. (So, right now, this patch doesn't add any tests, because there is
no behavior change.)

* Source/WebCore/platform/graphics/Font.cpp:
(WebCore::Font::systemFallbackFontForCharacter const):
* Source/WebCore/platform/graphics/Font.h:
* Source/WebCore/platform/graphics/FontCache.h:
* Source/WebCore/platform/graphics/FontCascadeFonts.cpp:
(WebCore::FontCascadeFonts::glyphDataForSystemFallback):
* Source/WebCore/platform/graphics/SystemFallbackFontCache.cpp:
(WebCore::SystemFallbackFontCache::systemFallbackFontForCharacter):
* Source/WebCore/platform/graphics/SystemFallbackFontCache.h:
(WebCore::add):
(WebCore::SystemFallbackFontCache::CharacterFallbackMapKeyHashTraits::constructDeletedValue):
(WebCore::SystemFallbackFontCache::CharacterFallbackMapKeyHashTraits::isDeletedValue):
* Source/WebCore/platform/graphics/cocoa/FontCacheCoreText.cpp:
(WebCore::lookupFallbackFont):
(WebCore::FontCache::systemFallbackForCharacters):
(WebCore::createFontForCharacters): Deleted.
* Source/WebCore/platform/graphics/coretext/FontCascadeCoreText.cpp:

Canonical link: https://commits.webkit.org/267335@main
